### PR TITLE
[None][feat] Enable MM encoder cache on Qwen3.x and Gemma4 VLMs

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Description

* Why?

Repeated or chunked multimodal requests re-ran encoder work on supported
Qwen3.x and Gemma4 models, increasing latency and GPU utilization.
Duplicated model-specific multimodal plumbing also made this behavior
harder to maintain consistently.

* What?

Refactor Gemma4 and Qwen3.x VLMs to reuse the multimodal base mixin and
shared encoder flow, removing duplicated multimodal boilerplate.

Enable cached encoder embeddings for supported models, preserve
model-specific behavior, exclude Qwen Image Bench from caching, and
document support.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Refactored Gemma4 and Qwen3.x VLM multimodal plumbing to reuse a shared `MultimodalModelMixin`/model base flow:
  - `Gemma4MultimodalModelBase` introduced and `Gemma4ForConditionalGeneration` updated to inherit it (centralizing multimodal encoding + LM integration hooks, and renaming token-id exposure to `multimodal_token_ids`).
  - `Gemma4UnifiedForConditionalGeneration` now inherits `Gemma4MultimodalModelBase`, removing the prior custom multimodal forward/fusion path and switching to encoder-free vision/audio feature extraction + shared multimodal fusion.
  - `Mistral3VLM` hook renamed/reshaped: `get_language_model_forward_kwargs` → `get_language_model_extra_forward_kwargs` to align with the shared extra-forward-kwargs pattern.
  - `Qwen3VLModelBase` forward rewritten to use shared multimodal lifecycle methods (`select_multimodal_params`, `after_active_multimodal_embeddings`, `_fuse_multimodal_embeddings`, `get_language_model_extra_forward_kwargs`) and to enforce the new multimodal token-id property name (`multimodal_token_ids`).
- Encoder-caching + cache prefetch behavior tightened/extended:
  - `MultimodalModelMixin.multimodal_token_ids` now defaults to `None` (instead of throwing), enabling the sentinel/OOV path in `fuse_input_embeds`.
  - `MultimodalModelMixin.select_multimodal_params` now filters params only when `param.has_content()` and the chunk has multimodal tokens (via `multimodal_runtime`/`num_mm_tokens_in_chunk` rules).
  - Cross-iteration encoder prefetch now propagates request-specific multimodal item ordering (`mm_item_order`) into `MultimodalParams`.
  - `PyTorchModelEngine._prepare_multimodal_indices` now derives `mm_token_ids` via the mixin-style `multimodal_token_ids` first, falling back to legacy `mm_token_ids`.
- Model-specific encoder-cache enablement:
  - `supports_encoder_cache = True` added for `_Qwen3_5VLModel` and `Qwen3MoeVLModel`.
  - Qwen Image Bench explicitly rebuilds multimodal config with `encoder_cache_max_bytes = 0` to exclude it from caching.
- Compatibility/error-handling changes to validate:
  - Gemma4 unified wrapper now hard-requires `config.image_token_id` (raises `ValueError` if missing).
  - Qwen3VLModelBase now directly requires `mm_encoder` for encoder-backed multimodal embedding; validates exactly one packed embedding tensor is returned.
  - Multimodal API call sites/tests updated to use `encode_multimodal_inputs` and shared multimodal selection helpers.
- Documentation:
  - `docs/source/models/supported-models.md` updated to enumerate all architectures implementing `MultimodalModelMixin` that support multimodal encoder side-stream + embeddings cache (replacing the previous single-model statement).

### Test changes (files under `tests/` touched)
- `tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py`
  - Added `_CacheStubModel` (cache-capable stub), `_make_cacheable_request(...)`.
  - Added CUDA tests:
    - `test_cross_iter_prefetch_populates_and_reuses_persistent_cache`
    - `test_cross_iter_prefetch_mixed_cache_hit_and_miss_encodes_only_miss`
    - `test_cross_iter_prefetch_cache_model_preserves_uncacheable_fallbacks`
    - `test_cross_iter_prefetch_does_not_synchronize_main_stream`
    - `test_cross_iter_prefetch_does_not_rewrite_request_local_embedding`
    - `test_prefetched_embedding_records_main_consumer_stream`
  - Added/updated CUDA-only dispatch materialization coverage via `@requires_cuda` for `test_cross_iter_prefetch_materializes_on_side_stream`.
- `tests/unittest/_torch/executor/test_pytorch_model_engine.py`
  - Added tests covering `_prepare_multimodal_indices` behavior for both:
    - models exposing `multimodal_token_ids`
    - legacy models exposing `mm_token_ids`
- `tests/unittest/_torch/modeling/test_gemma4_multimodal.py`
  - Updated multimodal API usage to call `encode_multimodal_inputs` (instead of `_forward_multimodal_encoder`).
  - Added `_Gemma4EncoderCacheHarness` and `test_encoder_cache_reuses_image_embedding_across_requests`.
  - Renamed/changed expectation for single-request multiple modalities:
    - `test_single_request_with_multiple_modalities_raises` (behavior update from prior “allowed” test).
  - Updated existing cache/order tests to use the new mixin methods/helpers.
- `tests/unittest/_torch/modeling/test_modeling_gemma4_unified.py`
  - Added `test_wrapper_rejects_missing_image_token_id`.

### Coverage / registration
- Whether these tests are present in `tests/integration/test_lists/` (test-db/qa entries) could not be confirmed from the provided change summary.
- Verdict: needs follow-up.

### CODE REVIEW

**Review by laikhtewari — APPROVED** [review]

**Review by arysef — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 1]
```diff
@@ -550,6 +545,245 @@ def call_with_text_prompt(
...
+        """Build Gemma4-specific language-model forward arguments."""
+        del position_ids, forward_kwargs
+        mm_token_type_ids = None
+        if raw_input_ids is not None and mm_inputs.input_ids is None:
+            mm_token_type_ids = torch.zeros_like(raw_input_ids, dtype=torch.long)
+            mm_token_type_ids[raw_input_ids == self.image_token_ids[0]] = 1
+            if self.video_token_ids is not None:
+                mm_token_type_ids[raw_input_ids == self.video_token_ids[0]] = 2
+            if self.audio_token_ids is not None:
+                mm_token_type_ids[raw_input_ids == self.audio_token_ids[0]] = 3
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:757]: Should no longer be an issue.

File: tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py [hunk 2]
```diff
@@ -259,3 +421,112 @@ def test_cross_iter_prefetch_skips_in_flight_and_cached(monkeypatch):
...
+            encoder_forward_fn=lambda _: pytest.fail("attached embeddings must skip the encoder"),
+            multimodal_params=[param],
+        )
+
+    consumer_stream.synchronize()
+    expected_calls = {
+        (embedding.data_ptr(), consumer_stream.cuda_stream) for embedding in embeddings
+    }
+    assert expected_calls.issubset(record_stream_calls)
+    torch.testing.assert_close(gathered[0], torch.ones(4 * embedding_count, device="cuda"))
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/multimodal/test_mm_encoder_cross_iter_prefetch.py]: `l0_h100.yml` should already include this via:

```yaml
- unittest/_torch/multimodal -k "not nemotron_nano_v2_vl_fp8"
```

Please verify if that's sufficient to ensure coverage.

File: tensorrt_llm/_torch/models/modeling_qwen_image_bench.py [hunk 3]
```diff
@@ -95,6 +97,16 @@ def load_weights(
...
+        if model_config.multimodal_config is not None:
+            model_config = replace(
+                model_config,
+                multimodal_config=MultimodalConfig(
+                    **model_config.multimodal_config.model_dump(
+                        exclude={"encoder_cache_max_bytes"}
+                    ),
+                    encoder_cache_max_bytes=0,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_qwen_image_bench.py:111]: This should no longer be an issue.

File: docs/source/models/supported-models.md
Comment (aswinvisva) [inline docs/source/models/supported-models.md]: nit: Should include Gemma4 & Qwen3?

Comment (2ez4bz) [reply docs/source/models/supported-models.md]: Came from the wrong commit.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -2995,8 +2995,11 @@ def get_padded_piecewise_tokens(tokens):
     def _prepare_multimodal_indices(self, input_ids: list[int]):
         input_ids = torch.tensor(input_ids, dtype=torch.int, device="cpu")
         vocab_size = self.model.config.vocab_size
-        # TODO: unify naming of mm_token_ids across models
-        mm_token_ids = getattr(self.model, "mm_token_ids", None)
+        # `multimodal_token_ids` is the common wrapper-model contract. Keep the legacy name as a
+        # fallback for models not yet migrated to `MultimodalModelMixin`.
+        mm_token_ids = getattr(self.model, "multimodal_token_ids", None)
```
Comment (aswinvisva) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:3063]: For models which don't implement `multimodal_token_ids` from the mixin, it will raise `NotImplementedError` 

```
@property
    def multimodal_token_ids(self) -> Optional[Sequence[int] | torch.Tensor]:
        """Return placeholder token ids in `input_ids` replaced by MM embeds.

        These are sentinel token positions whose text embeddings are replaced
        by multimodal embeddings. Return `None` to use the out-of-vocabulary
        sentinel behavior in `fuse_input_embeds`.
        """
        raise NotImplementedError
```

Should this instead be wrapped by a try/except?

Comment (2ez4bz) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:3063]: Good catch. Switched the base mixin to return `None` instead.

File: tensorrt_llm/_torch/models/modeling_qwen3vl.py [hunk 5]
```diff
@@ -1459,138 +1468,137 @@ def split_mm_embeds(self, mm_embed, deepstack_num_level):
...
-            # E/P prefill: encoder already ran; use attached embeddings.
-            else:
-                mm_embeds = get_attached_multimodal_embeddings(mm_multimodal_params)
-            mm_embeds = find_input_mm_embeds(mm_embeds, mm_multimodal_params)
+        if not has_raw_image_or_video_data and not getattr(self, "support_mm_disagg", False):
+            raise NotImplementedError(
+                f"{type(self)} does not support disaggregated inference yet. Please unset "
+                "the TLLM_MULTIMODAL_DISAGGREGATED environment variable, or set it to '0'."
+            )
+        return multimodal_params
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_qwen3vl.py:1423]: All registered Qwen3-VL-derived wrappers set `support_mm_disagg=True` via `@support_multimodal_disaggregated`, so attached-only cache/prefetch embeddings are accepted regardless of whether disaggregated mode is enabled.

File: tensorrt_llm/_torch/models/modeling_qwen_image_bench.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_qwen_image_bench.py:111]: Added.

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 6]
```diff
@@ -807,99 +807,29 @@ def draft_config(self):
...
+        del raw_input_ids, position_ids, mm_inputs, forward_kwargs
         return {
-            "attn_metadata": attn_metadata,
-            "input_ids": input_ids,
-            "position_ids": position_ids,
-            "inputs_embeds": inputs_embeds,
-            "return_context_logits": return_context_logits,
             "spec_metadata": spec_metadata,
             "resource_manager": resource_manager,
         }
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py:870]: ADded `lora_params` to mistral.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:757]: `image_token_id` is required for supported Gemma4 Unified configs; initialization already assumes it is present when constructing `_mm_token_ids`. An image-less config is unsupported, so this dereference is not reachable for a valid model config.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 7]
```diff
@@ -550,6 +545,245 @@ def call_with_text_prompt(
...
+        if audio_features_list and self.embed_audio is not None:
+            per_audio_embeddings = [
+                self._get_audio_features(audio_features, audio_features_mask_list[index])
+                for index, audio_features in enumerate(audio_features_list)
+            ]
+            multimodal_embeddings.append(torch.cat(per_audio_embeddings, dim=0))
+
+        if not multimodal_embeddings:
+            raise ValueError("Gemma4 received active multimodal parameters without encoder inputs")
+        return torch.cat(multimodal_embeddings, dim=0)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:704]: Added a check that raises if a single request contains multiple modalities.

File: docs/source/models/supported-models.md [hunk 8]
```diff
@@ -127,11 +127,17 @@ Note:
...
 
 | Model Architecture | Multimodal Encoder Side Stream | Multimodal Embeddings Cache |
 | ------------------ | ------------------------------ | --------------------------- |
+| `Gemma4ForConditionalGeneration` | Yes | Yes |
+| `Gemma4UnifiedForConditionalGeneration` | Yes | Yes |
 | `Mistral3ForConditionalGeneration` | Yes | Yes |
+| `Qwen3VLForConditionalGeneration` | Yes | Yes |
+| `Qwen3VLMoeForConditionalGeneration` | Yes | Yes |
+| `Qwen3_5ForConditionalGeneration` | Yes | Yes |
+| `Qwen3_5MoeForConditionalGeneration` | Yes | Yes |
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply docs/source/models/supported-models.md:143]: TODO: Revisit this once #16554 is merged; that PR should bridge this documentation gap.

File: tensorrt_llm/_torch/models/modeling_mistral.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py:870]: - The Mistral concern is fixed: `lora_params`, `spec_metadata`, and `resource_manager` are forwarded.
- Gemma4 forwards the only relevant extra argument, `lora_params`.
- Gemma4’s language model does not consume `spec_metadata` or `resource_manager`; forwarding them would merely pass unused kwargs through its layers.

Please verify.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 9]
```diff
@@ -550,6 +545,250 @@ def call_with_text_prompt(
...
+        """Build Gemma4-specific language-model forward arguments."""
+        del position_ids, forward_kwargs
+        mm_token_type_ids = None
+        if raw_input_ids is not None and mm_inputs.input_ids is None:
+            mm_token_type_ids = torch.zeros_like(raw_input_ids, dtype=torch.long)
+            mm_token_type_ids[raw_input_ids == self.image_token_ids[0]] = 1
+            if self.video_token_ids is not None:
+                mm_token_type_ids[raw_input_ids == self.video_token_ids[0]] = 2
+            if self.audio_token_ids is not None:
+                mm_token_type_ids[raw_input_ids == self.audio_token_ids[0]] = 3
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:757]: Added a check that it is not `None` in the constructor.

**Review by aswinvisva — APPROVED** [review]
lgtm

**Review by Tabrizian — APPROVED** [review]

File: docs/source/models/supported-models.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (2ez4bz) [reply docs/source/models/supported-models.md:143]: Will address in a followup PR.
